mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -244,23 +244,33 @@ liftWriteResult :: WriteResult -> AWS ()
 liftWriteResult = \case
   WriteOk ->
     pure ()
-  WriteDestinationExists a ->
+  WriteKo (WriteDestinationExists a) ->
     throwM $ DestinationAlreadyExists a
+  WriteKo (WriteInaccessibleBucket b) ->
+    throwM $ InaccessibleBucket b
 
 writeWithMode :: WriteMode -> Address -> Text -> AWS WriteResult
-writeWithMode w a t = eitherT pure (const $ pure WriteOk) $ do
-  case w of
-    Fail -> whenM (lift $ exists a) . left $ WriteDestinationExists a
-    Overwrite -> return ()
-  void . lift . send $ fencode' putObject a (toBody . T.encodeUtf8 $ t) & poServerSideEncryption .~ Just sse
+writeWithMode w a t = do
+  check <- checkWriteWithMode w a
+  case check of
+    Nothing -> eitherT pure (const $ pure WriteOk) (void . lift . send $ fencode' putObject a (toBody . T.encodeUtf8 $ t) & poServerSideEncryption .~ Just sse)
+    Just e  -> pure (WriteKo e)
```

Comment:
Potential race condition here? Shouldn't we do it atomically? (handling error in `putObject` call)